src/Chapter3.hs

```diff
@@ -480,6 +528,60 @@ After defining the city, implement the following functions:
    and at least 10 living __people__ inside in all houses of the city totally.
 -}
 
+data Castle = Castle String 
+    deriving (Show, Eq)
+
+data Wall = Wall Bool 
+    deriving (Show, Eq)
+
+data Community = Church | Library 
+    deriving (Show, Eq)
+
+data Housing = Housing
+    { oneOccupant    :: Int
+    , twoOccupants   :: Int
+    , threeOccupants :: Int
+    , fourOccupants  :: Int
+    } deriving (Show, Eq)
+
+data City = City
+    { castle    :: Maybe Castle
+    , wall      :: Wall
+    , community :: Community
+    , housing   :: Housing
+    } deriving (Show, Eq)
+
+buildCastle :: City -> City
+buildCastle city = 
+    city { castle = Just (Castle "NewCastle") }
+
+buildHouse :: Housing -> City -> City
+buildHouse newHouses city =
+    city { housing = Housing
+            { oneOccupant    = (oneOccupant newHouses) + (oneOccupant $ housing city)
+            , twoOccupants   = (twoOccupants newHouses) + (twoOccupants $ housing city)
+            , threeOccupants = (threeOccupants newHouses) + (threeOccupants $ housing city)
+            , fourOccupants  = (fourOccupants newHouses) + (fourOccupants $ housing city)
+            }
+         }
+
+buildWalls :: City -> City
+buildWalls city = 
+    if isJust (castle city) then 
+        if totalHouses (housing city) >= 10 then 
+            city { wall = Wall True }
+        else city
+    else city
+    where 
+        totalHouses :: Housing -> Int
+        totalHouses houses = 
+            (oneOccupant $ houses) + (twoOccupants $ houses) + (threeOccupants $ houses) + (fourOccupants $ houses)
+
+        isJust :: Maybe Castle -> Bool
+        isJust (Just _) = True
+        isJust _        = False
```

Comment:
You might be interested, but this is actually a standard function :smile: 

* http://hackage.haskell.org/package/base-4.14.0.0/docs/Data-Maybe.html#v:isJust

Though, if you model your types differently, you won't need to use `Maybe` separately